lib/uefi/blockio2_protocols.cpp

```diff
@@ -0,0 +1,123 @@
+
+#include "blockio2_protocols.h"
+
+#include <kernel/thread.h>
+#include <kernel/vm.h>
+#include <lib/bio.h>
+#include <lk/err.h>
+#include <lk/trace.h>
+#include <stddef.h>
+#include <stdio.h>
+#include <uefi/protocols/block_io2_protocol.h>
+#include <uefi/types.h>
+
+#include "events.h"
+#include "io_stack.h"
+#include "memory_protocols.h"
+#include "switch_stack.h"
+#include "thread_utils.h"
+
+#define LOCAL_TRACE 0
+
+namespace {
+
+struct EfiBlockIo2Interface {
+  EfiBlockIo2Protocol protocol;
+  EfiBlockIoMedia media;
+  void* dev;
+};
+
+EfiStatus reset(EfiBlockIo2Protocol* self, bool extended_verification) {
+  return UNSUPPORTED;
+}
+
+// Read from dev, after I/O completes, signal token->event and set
+// token->transaction_status
+EfiStatus read_blocks_async(bdev_t* dev, uint64_t lba, EfiBlockIo2Token* token,
+                            size_t buffer_size, void* buffer) {
+  if (lba >= dev->block_count) {
+    printf("OOB async read %s %llu %u\n", dev->name, lba, dev->block_count);
+    return END_OF_MEDIA;
+  }
+  if (token == nullptr) {
+    printf("Invalid token %p\n", token);
+    return INVALID_PARAMETER;
+  }
+  // First draft of this API will just use a background thread.
+  // More efficient version can be implemented once LK's bio layer
+  // supports async IO
+  auto thread = thread_create(
+      "async_bio",
+      [dev, lba, buffer_size, buffer, token]() {
+        auto aspace = set_boot_aspace();
+        vmm_set_active_aspace(aspace);
+        auto bytes_read =
+            bio_read_block(dev, buffer, lba, buffer_size / dev->block_size);
+        if (static_cast<size_t>(bytes_read) != buffer_size) {
+          token->transaction_status = DEVICE_ERROR;
+        } else {
+          token->transaction_status = SUCCESS;
+        }
+        signal_event(token->event);
+        return 0;
+      },
+      get_current_thread()->priority, kIoStackSize);
+  if (thread == nullptr) {
+    printf("Failed to create thread for IO read\n");
+    return DEVICE_ERROR;
+  }
+
+  auto err = thread_resume(thread);
```

Comment:
If the thread isn't detached, then the remants of the thread will stick around forever until something calls thread_join on it. If the intention is for someone to poll for transaction_status to be SUCCESS, then you should detach it here.